### ISSUE TITLE
Is unsetting or skipping parameters in Breadcrumbs::register possible?

### ISSUE THREAD

**Issue by Inuyaki** [davejamesmiller/laravel-breadcrumbs#151]
First of all, great little tool you got there, really well designed!

Our page is divided into seperate areas for our clients, but all of them have access to the same page. They wanted to have that reflected in the url, so we build a little prefixing group like:
   `Route::prefix('{clientident}')->middleware('clientident')->group(function() { all the routes });`

To not be troubled with that stupid clientident parameter (which we sadly need for our middleware to check if that client is indeed a client) in every single controller method, we did override the Controller callAction with
    `if (isset($parameters['clientident'])) { unset($parameters['clientident']); }`

But we still get that pesky parameter in all our Breadcrumbs, which is a little bit cumbersome and kinda confusing to have a $_ in every single register.

Is there an easy way to skip a parameter before it gets injected?

Thanks for you help :)

**Comment by davejamesmiller** [comment]
You could call `Breadcrumbs::setCurrentRoute()` from your middleware or callAction method perhaps...

https://github.com/davejamesmiller/laravel-breadcrumbs#overriding-the-current-route

Or you could make a subclass of `BreadcrumbsManager` and override `getCurrentRoute()`...

https://github.com/davejamesmiller/laravel-breadcrumbs/blob/830085fa364f6c1adf7c624520824163acc8576b/src/BreadcrumbsManager.php#L241-L270

To do that you'd need to add this line to a service provider (e.g. `AppServiceProvider`) so Laravel uses the custom subclass instead of the original:

```php
use App\CustomBreadcrumbsManager;
use DaveJamesMiller\Breadcrumbs\BreadcrumbsManager;

//...
    public function register()
    {
        //...
        $this->app->singleton(BreadcrumbsManager::class, CustomBreadcrumbsManager::class);
    }
//...
```

**Comment by Inuyaki** [comment]
I went the subclass way, and it works, I just had to override the $params set at the end to
`$params = array_values(array_filter($route->parameters(), function($key) { return $key != 'clientident'; }, ARRAY_FILTER_USE_KEY));`

There is still something though, I can't override your singleton in AppServiceProvider because your ServiceProvider seems to be called later. When I write the line from above in your BreadcrumbsServiceProvider, it overrides your original singleton. In AppServiceProvider it does not work.
So my conclusion was that your ServiceProvider is called later, I don't actually know the chain here.
Any idea?

**Comment by davejamesmiller** [comment]
Hmm... It could be one of two things...

1. In Laravel 5.4 you could change the service provider load order in `config/app.php`, but Laravel 5.5 auto-loads them - my guess is they're added at the end. You could try manually adding mine above yours. You may also need to add mine to the [`dont-discover` array](https://medium.com/@taylorotwell/package-auto-discovery-in-laravel-5-5-ea9e3ab20518) in your composer.json - or it may be smart enough not to try to load it twice.

2. Or it could be because my service provider has `$defer = true` set, which means it's only loaded when it's first needed (e.g. when you call `Breadcrumbs::render()`). You may be able to force it to be eager loaded instead - I'm not sure though since that might also trigger loading the BreadcrumbsManager class...

Alternatively (in either case) you could subclass BreadcrumbsServiceProvider, override the register() method, then add the subclass to your `config/app.php` and add the original to `dont-discover`.

Seems a little convoluted though... so I'd consider allowing the Manager and Generator classes to be set in the config file, if you wanted to try making a PR for that.

**Comment by davejamesmiller** [comment]
On reflection I decided allowing the classes to be overwritten would be a useful feature anyway, so I've added it and released 4.2.0. Try it out and let me know if it works for you.

https://github.com/davejamesmiller/laravel-breadcrumbs#advanced-customisations

**Comment by Inuyaki** [comment]
Damn, you beat me to it... had no time until now ;)
Yeah, looks good (at least I would have done it the same) and works, so thanks for your help 👍
